filament/src/materials/dof/dofDownsample.mat

```diff
@@ -50,7 +50,21 @@ vertex {
     void dummy(){}
 
     void postProcessVertex(inout PostProcessVertexInputs postProcess) {
-        postProcess.vertex.xy = (postProcess.normalizedUV * materialParams.uvscale.xy * 2.0 - 0.5) * materialParams.uvscale.zw;
+        // It works to fix-up the UV *before* applying the offsets because materialParams.texelSize
+        // corresponds to one texel.
+        // on GL:           v0 = v - 0.5 * texel                                                (1)
+        //                  v1 = v - 0.5 * texel + 1 * texel
+        //                     = v + 0.5 * texel                                                (2)
+        // on MTL/VK:       v0 = 1 - ((1 - v) - 0.5 * texel)
+        //                     = v + 0.5 * texel                                                (3)
+        //                  v1 = 1 - ((1 - v) - 0.5 * texel + 1 * texel)
+        //                     = v - 0.5 * texel                                                (4)
+        // Because the all the operations on the 4 texels in the fragment shader don't depend on
```

Comment:
Can you elaborate on this? It seems that `outColor` depends on `w` which depends on `hdr`, which depends on the ordering of the texels in `s00`, `s01`, etc.

Reply:
Thank you for checking this -- I checked again and I think it is good.

So I think it works because the ordering of `s00`, etc... matches the ordering of `d[]` so, in turn, `c` follows the ordering of `d[]` and then `w` does too.

I don't think `hdr` "depends on hdr, which depends on the ordering of the texels in s00, s01, etc.". i.e. `hdr[0]` always depend on `s01`.


Another way to think of it is that we're "replacing" 4 texels by a single one. And those 4 texels are all interchangeable, meaning they don't depend on their neighbor. The downsampling uses the `CoC`, but it too is fetched in the same order (`d[]`).

What do you think?

Reply:
Okay yeah, that makes sense. I think I interpreted your comment meaning `s00` might refer to a different texel depending on which backend was used. But after examining it again, I don't think that's the case, correct?

Reply:
No that's the case! Depending on the backend `s01` and `s00` are swapped and `s11` and `s10` are swapped. But I claim that it doesn't change the final result because `d[]` is swapped in the same way. And `hdr` only works on a single value (it does the max(r,g,b)).

Reply:
👍